File: system-test/dotnet/Dsl.Core/UseCase/Shop/UseCases/BrowseCouponsVerification.cs
using Dsl.Core.Shared;
using Driver.Port.Shop.Dtos;
using Shouldly;

namespace Dsl.Core.Shop.UseCases;

public class BrowseCouponsVerification : ResponseVerification<BrowseCouponsResponse>
{
    public BrowseCouponsVerification(BrowseCouponsResponse response, UseCaseContext context)
        : base(response, context)
    {
    }

    public BrowseCouponsVerification ContainsCouponWithCode(string expectedCode)
    {
        Response.Coupons.ShouldContain(
            c => c.Code == expectedCode,
            $"Expected coupon with code '{expectedCode}' to be present, but was not found");
        return this;
    }

    public BrowseCouponsVerification CouponCount(int expectedCount)
    {
        Response.Coupons.Count.ShouldBe(expectedCount,
            $"Expected {expectedCount} coupons, but found {Response.Coupons.Count}");
        return this;
    }
}

File: system-test/java/src/main/java/com/optivem/shop/dsl/core/usecase/shop/usecases/BrowseCouponsVerification.java
package com.optivem.shop.dsl.core.usecase.shop.usecases;

import com.optivem.shop.dsl.core.shared.ResponseVerification;
import com.optivem.shop.dsl.core.shared.UseCaseContext;
import com.optivem.shop.dsl.driver.port.shop.dtos.BrowseCouponsResponse;

import static org.assertj.core.api.Assertions.assertThat;

public class BrowseCouponsVerification extends ResponseVerification<BrowseCouponsResponse> {
    public BrowseCouponsVerification(BrowseCouponsResponse response, UseCaseContext context) {
        super(response, context);
    }

    public BrowseCouponsVerification containsCouponWithCode(String expectedCode) {
        var coupons = getResponse().getCoupons();
        assertThat(coupons)
                .withFailMessage("Expected coupon with code '%s' to be present, but was not found", expectedCode)
                .anyMatch(c -> expectedCode.equals(c.getCode()));
        return this;
    }

    public BrowseCouponsVerification couponCount(int expectedCount) {
        var actualCount = getResponse().getCoupons().size();
        assertThat(actualCount)
                .withFailMessage("Expected %d coupons, but found %d", expectedCount, actualCount)
                .isEqualTo(expectedCount);
        return this;
    }
}
